src/index.js

```diff
@@ -30,39 +30,44 @@ class ESLintWebpackPlugin {
       files: parseFiles(this.options.files, this.getContext(compiler)),
       extensions: arrify(this.options.extensions),
     };
+    Object.freeze(options);
 
-    if (options.lintDirtyModulesOnly) {
-      const dirtyFileWatcher = new DirtyFileWatcher(
-        options.files,
-        options.extensions
-      );
+    /**
+     * @param {Compiler} runCompiler
+     */
+    const run = async (runCompiler) => {
+      /** @type {string[]} */
+      const files = [];
+      const wanted = parseFoldersToGlobs(options.files, options.extensions);
 
-      /* istanbul ignore next */
-      compiler.hooks.watchRun.tapPromise(
+      runCompiler.hooks.compilation.tap(
         'ESLintWebpackPlugin',
-        async (runCompiler) => {
-          const files = dirtyFileWatcher.getDirtyFiles(
-            runCompiler.fileTimestamps
-          );
+        (compilation) => {
+          compilation.hooks.succeedModule.tap(
+            'ESLintWebpackPlugin',
+            (module) => {
+              // @ts-ignore
+              // eslint-disable-next-line no-underscore-dangle
+              const file = module.resource;
 
-          if (files.length > 0) {
-            await linter({ ...options, files }, runCompiler);
-          }
+              if (file && micromatch.isMatch(file, wanted)) {
+                // Queue file for linting.
+                files.push(file);
+              }
+            }
+          );
         }
       );
-    } else {
-      compiler.hooks.run.tapPromise('ESLintWebpackPlugin', (runCompiler) => {
-        return linter(options, runCompiler);
-      });
 
-      /* istanbul ignore next */
-      compiler.hooks.watchRun.tapPromise(
-        'ESLintWebpackPlugin',
-        (runCompiler) => {
-          return linter(options, runCompiler);
-        }
+      runCompiler.hooks.emit.tapPromise('ESLintWebpackPlugin', async () =>
+        linter(options, files, runCompiler)
       );
+    };
+
+    if (!options.lintDirtyModulesOnly) {
+      compiler.hooks.run.tapPromise('ESLintWebpackPlugin', run);
     }
+    compiler.hooks.watchRun.tapPromise('ESLintWebpackPlugin', run);
```

Comment:
We can look at `compiler.watching` (webpack@5) to detect watching and improve DX

Reply:
should I test for that existence or not support v4?

Reply:
Ideally we should test it because we support v4, but manually tests are enough here

Reply:
I meant at runtime, detect it exists. I'm trying to find documentation on how to use `compiler.watching`. Any pointers?

Reply:
`compiler.watching` is not documented yes, but you can use it for getting `watching` or determinate what we are in watch

Reply:
I'm not as familiar with this api. How would I change this line to use it?

Reply:
Is there an example of this?

Reply:
Are you suggesting we do not tap into watchRun unless we are watching?

Reply:
like this?
```js
    if (compiler.watchMode ?? compiler.watching) {
      compiler.hooks.watchRun.tapPromise(ESLINT_PLUGIN, run);
    }
```